aws/templates/solution/solution_userpool.ftl

```diff
@@ -1,209 +1,301 @@
-    [#assign identityPoolName = formatIdentityPoolName(tier, component)]
-    [#assign userPoolClientName = formatUserPoolClientName(tier, component) ]
-    [#assign dependencies = [] ]
-    [#assign smsVerification = false]
-    [#assign schema = [] ]
-
-    [@cfDebug listMode appSettingsObject false /]
-
-    [#assign emailVerificationMessage = ""]
-    [#if (appSettingsObject.UserPool.EmailVerificationMessage)?has_content ]
-        [#assign emailVerificationMessage = appSettingsObject.UserPool.EmailVerificationMessage ]
-    [/#if]
-
-    [#assign emailVerificationSubject = ""]
-    [#if (appSettingsObject.UserPool.EmailVerificationSubject)?has_content ]
-        [#assign emailVerificationSubject = appSettingsObject.UserPool.EmailVerificationSubject ]
-    [/#if]
-
-    [#assign smsVerificationMessage = ""]
-    [#if (appSettingsObject.UserPool.SMSVerificationMessage)?has_content ]
-        [#assign emailVerificationSubject = appSettingsObject.UserPool.SMSVerificationMessage ]
-    [/#if]
-
-    [#assign emailInviteMessage = "" ]
-    [#if (appSettingsObject.UserPool.EmailInviteMessage)?has_content ]
-        [#assign emailInviteMessage = appSettingsObject.UserPool.EmailInviteMessage ]
-    [/#if]
-
-    [#assign emailInviteSubject = "" ]
-    [#if (appSettingsObject.UserPool.EmailInviteSubject)?has_content ]
-        [#assign emailInviteSubject = appSettingsObject.UserPool.EmailInviteSubject ]
-    [/#if]
-
-    [#assign smsInviteMessage = ""] 
-    [#if (appSettingsObject.UserPool.SMSInviteMessage)?has_content ]
-        [#assign smsInviteMessage = appSettingsObject.UserPool.SMSInviteMessage ]
-    [/#if]
-
-    [#if (userpool.MFA?has_content && userpool.MFA) || (userpool.VerifyPhone?has_content && userpool.VerifyPhone)]
-        [#if deploymentSubsetRequired("iam", true) &&
-        isPartOfCurrentDeploymentUnit(userPoolId) ]
-
-            [@createRole
-                mode=listMode
-                id=userPoolRoleId
-                trustedServices=["cognito-idp.amazonaws.com" ]
-                policies=
-                    [
-                        getPolicyDocument(
-                            snsPublishPermission(),
-                            "smsVerification" 
-                        )
-                    ]
-            /]
-
-            [#assign phoneSchema = getUserPoolSchemaObject( 
-                                        "phone_number",
-                                        "String",
-                                        true,
-                                        true)]
-            [#assign schema = schema + [ phoneSchema ]]
-
-            )]
+    [#list requiredOccurrences(
+        getOccurrences(tier, component),
+        deploymentUnit) as occurrence]
 
+        [@cfDebug listMode occurrence false /]
+
+        [#assign core = occurrence.Core ]
+        [#assign configuration = occurrence.Configuration ]
+
+        [#assign userPoolId = formatUserPoolId(tier, component, occurrence)]
```

Comment:
Add the Id and Name functions could be simplified to just accept  (occurrence extension...) couldn't they? See lambda and s3 for examples.

Means now having pass around tier and component everywhere